solver/pb/ops.proto

```diff
@@ -211,4 +199,67 @@ message Definition {
 message HostIP {
 	string Host = 1;
 	string IP = 2;
+}
+
+message FileOp {
+	repeated FileAction actions = 2;
+}
+
+message FileAction {
+	int64 input = 1 [(gogoproto.customtype) = "InputIndex", (gogoproto.nullable) = false]; // could be real input or target (target index + max input index)
+	int64 secondaryInput = 2 [(gogoproto.customtype) = "InputIndex", (gogoproto.nullable) = false]; // --//--
+  int64 output = 3 [(gogoproto.customtype) = "OutputIndex", (gogoproto.nullable) = false];
+	oneof action {
+		FileActionCopy copy = 4;
+		FileActionMkFile mkfile = 5;
+		FileActionMkDir mkdir = 6;
+		FileActionRm rm = 7;
+	}
+}
+
+message FileActionCopy {
+	string src = 1;
+	string dest = 2;
+	ChownOpt owner = 4;
+	int32 mode = 5;
+	bool followSymlink = 6;
+	bool dirCopyContents = 7;
+	bool attemptUnpack = 8;
+	bool createDestPath = 9;
+	bool allowWildcard = 10;
+	bool allowEmptyWildcard = 11;
+	int64 timestamp = 12;
+}
+
+message FileActionMkFile {
+	string path = 1;
+	int32 mode = 2;
+	bytes data = 3;
```

Comment:
I think this is the first construct within the LLB which has the potential to create very large protobuf in one go in a possibly unconstrained way. Everything else can make huge graphs just through composition but is fairly limited by the complexity of the most complex graph you might sanely build.

That seems pretty unavoidable but has implications elsewhere (e.g. on maximum grpc size).

Reply:
> unavoidable...

Or maybe... could we use the session (along the lines of fssync) to somehow provide the data in bulk oob from the llb graph itself?

Reply:
I think this can be done with docs or we can define limits. This is not an alternative for sources but should be only used for small one-off files. Eg. https://github.com/moby/moby/issues/34423 is a good candidate.